lib/pbench/server/cache_manager.py

```diff
@@ -799,6 +800,131 @@ def extract(tarball_path: Path, path: str) -> Inventory:
             tarball_path, f"Unexpected error from {tar_path}: {error_text!r}"
         )
 
+    def get_contents(self, path: str, origin: str) -> JSONOBJECT:
+        """Return a description of a directory.
+
+        Args:
+            path: relative path within the tarball
+            origin: root URI path for the dataset
+
+        Returns:
+            A "json" dict describing the target.
+        """
+
+        with LockManager(self.lock) as lock:
+            artifact: Path = self.get_results(lock) / path
+            try:
+                # NOTE: os.path.abspath() removes ".." but Path.absolute(),
+                # doesn't, meaning the latter allows a query with path "..",
+                # and we don't want to allow reaching outside of the tarball.
+                arel = Path(os.path.abspath(artifact)).relative_to(self.unpacked)
+            except Exception:
+                raise CacheExtractError(self.name, path)
+            if artifact.is_dir() and not artifact.is_symlink():
+                dir_list = []
+                file_list = []
+                for f in artifact.iterdir():
+                    relative = f.relative_to(self.unpacked)
+                    append_to = file_list
+                    if f.is_symlink():
+                        target = f.resolve()
+                        try:
+                            link = target.relative_to(self.unpacked)
+                        except Exception:
+                            link = f.readlink()
+                            uri = f"{origin}/inventory/{relative}"
+                            type = CacheType.BROKEN
+                        else:
+                            if target.is_dir():
+                                uri = f"{origin}/contents/{link}"
+                                type = CacheType.DIRECTORY
+                                append_to = dir_list
+                            elif target.is_file():
+                                uri = f"{origin}/inventory/{link}"
+                                type = CacheType.FILE
+                            else:
+                                uri = f"{origin}/inventory/{relative}"
+                                type = CacheType.OTHER
```

Comment:
I concur on the `BROKEN` case...but why is that appropriate for the `OTHER` case?  (In the `OTHER` case, we successfully resolved the symlink to a directory entry accessible within the tarball via a relative path, so won't `f"{origin}/inventory/{link}"` and `f"{origin}/inventory/{relative}" address the same file, with the latter just taking a more circuitous route, if the route is different at all?)

Reply:
That's actually a comment for `inventory` rather than `contents`. We don't want to expose the real target of an `OTHER` file (e.g., a fifo or whatever), which is why we use `OTHER`. Yeah, it's possible that `inventory` will transparently resolve the link and return the `fifo`, which I suspect we really don't want. I'm not going to mess with that here.